### PR TITLE
Event Hook Decorators, Better AudioTrack, EQ & Copyright Year Fix, + more

### PR DESCRIPTION
**Pull request by SoulSen**
### Checks and guidelines:

* [x] Have you checked that there aren't other open [Pull Requests](../../pulls) for the same update/change?
* [x] Have you checked that only **single quotes** are used in the code, apart from the doc-strings?
* [x] Have you run a lint program on your code prior to submission?
* [x] Have you checked if `python run_tests.py` returns no errors?

Tests output: 
```
-- flake8 test --
OK
-- pylint test --
OK
```

### Type of change

* [x] Bug fix
* [x] New feature
* [x] Improvement
* [x] Breaking change
* [x] This change is a documentation update

### Describe the changes:

- Added Event Hook Decorators, changed `Client._event_hooks` to a dictionary for better compatibility with Decorators. `None` key are all the unknown event hooks, meaning we don't know what even they belong to, but the keys with `Event` are the keys we know what events they belong to. This is more efficient so we don't have to dispatch every single function. 

- Added `connect_back` parameter to `Client`, allowing users to be able to have the player connect to their original node that they connected to from it's current node. 

- Removed `AudioTrack.preferences` and replaced with `AudioTrack.extras`, this is now a dictionary full of extra info. about the `AudioTrack`. 

- Added `__slots__` for `Stats` and `Penalty`, for better memory stuff. 

- Fixed EQ Bands conditional in `DefaultPlayer.set_gains()` which can never be true, to check if it's an invalid band and if so it will raise `InvalidBand`, which is a new exception added. 

- Added and documented `end_time` and `no_replace` parameters in `DefaultPlayer.play()`. `end_time` defaults to -1 and is changed to `AudioTrack.duration`, if it is set, it's checked if it's less than 0 or greater than `AudioTrack.duration`. `no_replace` is also defaulted as `False`.
  
- Fixed Copyright Year from 2018 -> 2019.

### CODE REVIEW

File: lavalink/client.py [hunk 1]
```diff
@@ -27,9 +27,6 @@ class Client:
         The user id of the bot.
     shard_count: Optional[int]
         The amount of shards your bot has.
-    pool_size: Optional[int]
```
Comment (AndyTempel) [inline lavalink/client.py]: Where did that go?

File: lavalink/client.py [hunk 2]
```diff
@@ -39,26 +36,44 @@ class Client:
...
+        :param hook:
+            The hook to be added, that will be dispatched when an event is dispatched.
+            If `event` parameter is left empty, then it will run when any event is dispatched.
+        :param event:
+            The event the hook belongs to. This will dispatch when that specific event is
+            dispatched.
+        """
+        event_hooks = self._event_hooks.get(event)
+        if hook not in event_hooks:
+            if event_hooks is None:
```
Comment (AndyTempel) [inline lavalink/client.py]: @Devoxin what's the more performant version of this?

File: lavalink/client.py [hunk 3]
```diff
@@ -195,6 +210,17 @@ def add_node(self, host: str, port: int, password: str, region: str,
         else:
             return
 
+    def on(self, event: Event):
+        def decorator(func):
+            event_hook = self._event_hooks.get(event)
```
Comment (AndyTempel) [inline lavalink/client.py]: Re-use `add_event_hook` method

File: lavalink/client.py [hunk 4]
```diff
@@ -203,11 +229,18 @@ def add_node(self, host: str, port: int, password: str, region: str,
         :param event:
             The event to dispatch to the hooks.
         """
-        for hook in self._event_hooks:
-            try:
+        unknown_events = self._event_hooks.get(None) or []
+        registered_events = self._event_hooks.get(event) or []
+
+        event_hooks = unknown_events + registered_events
```
Comment (AndyTempel) [inline lavalink/client.py]: See below

File: lavalink/client.py [hunk 5]
```diff
@@ -203,11 +229,18 @@ def add_node(self, host: str, port: int, password: str, region: str,
...
+
+        try:
+            for hook in event_hooks:
```
Comment (AndyTempel) [inline lavalink/client.py]: Better was so we don't need to make two shallow copies of both lists and just use itertools:
`for item in itertools.chain(unknown_events, registered_events):`

File: lavalink/client.py [hunk 6]
```diff
@@ -203,11 +229,18 @@ def add_node(self, host: str, port: int, password: str, region: str,
...
                 if inspect.iscoroutinefunction(hook):
                     await hook(event)
```
Comment (AndyTempel) [inline lavalink/client.py]: `asyncio.ensure_future()`

File: lavalink/client.py [hunk 7]
```diff
@@ -203,11 +229,18 @@ def add_node(self, host: str, port: int, password: str, region: str,
...
                 else:
                     hook(event)
```
Comment (AndyTempel) [inline lavalink/client.py]: @Devoxin @Zenrac should we really allow non-async funcs in here? And potentially holdup the whole thread and WS conn?

File: lavalink/models.py [hunk 8]
```diff
@@ -16,16 +17,15 @@ class TrackNotBuilt(Exception):
...
+                 'extra')
 
-    def __init__(self, requester, **kwargs):
+    def __init__(self, requester):
         self.requester = requester
-        self.preferences = kwargs
 
     @classmethod
-    def build(cls, track, requester, **kwargs):
+    def build(cls, track, requester, extra: dict = None):
```
Comment (AndyTempel) [inline lavalink/models.py]: Isn't replacing kwargs with dict making things less readable? kwargs gets converted to dict anyway.

File: lavalink/models.py [hunk 9]
```diff
@@ -35,6 +35,7 @@ def build(cls, track, requester, **kwargs):
             new_track.stream = track['info']['isStream']
             new_track.title = track['info']['title']
             new_track.uri = track['info']['uri']
+            new_track.extra = extra or {}
```
Comment (AndyTempel) [inline lavalink/models.py]: I don't get why is this here and not like it was. Seems like over-complicating

File: lavalink/client.py (hunk 1 above)
Comment (SoulSen) [reply lavalink/client.py]: Devoxin removed it, so I don't know.

File: lavalink/client.py (hunk 7 above)
Comment (SoulSen) [reply lavalink/client.py]: It was non-async before so I kept it like that

Comment (AndyTempel) [reply lavalink/client.py]: ~~ya know that's why i mention those two~~

File: lavalink/models.py (hunk 8 above)
Comment (SoulSen) [reply lavalink/models.py]: I'm not sure, Devoxin wanted it to be a parameter that's a dict rather than kwargs

File: lavalink/models.py (hunk 9 above)
Comment (SoulSen) [reply lavalink/models.py]: Since I defaulted it as `None` it would either be `{}` or if it's set then the actual dict.

Comment (AndyTempel) [reply lavalink/models.py]: or use `__init__` and have self.extra = extra
and if i'm not mistaken kwargs returns empty dict

File: lavalink/models.py (hunk 8 above)
Comment (AndyTempel) [reply lavalink/models.py]: @Devoxin pls explain

Comment (Devoxin) [reply lavalink/models.py]: My reasoning for the above is having stuff like `AudioTrack.build(..., spotify_playlist='blah', track_index=32)` is harder for people to interpret what the line does if looking at the code from an open-source perspective. Passing a dict instead like `AudioTrack.build(..., {'spotify_playlist': 'blah', 'track_index': 32})` looks clearer in portraying that it's a dictionary of extra information to associate with the track. Lastly, kwargs gets converted to a dict anyway as you say, which there's no point in wasting CPU cycles having the interpreter/compiler do this for us when the user is capable of doing it themselves.

File: lavalink/nodemanager.py [hunk 10]
```diff
@@ -53,6 +53,8 @@ def add_node(self, host: str, port: int, password: str, region: str, name: str =
         node = Node(self, host, port, password, region, name, resume_key, resume_timeout)
         self.nodes.append(node)
 
+        log.info("[NODE-{}] Successfully added Node to Node Manager".format(node.name))
```
Comment (Devoxin) [inline lavalink/nodemanager.py]: Nit: `Successfully added node.`

File: lavalink/nodemanager.py [hunk 11]
```diff
@@ -61,6 +63,7 @@ def remove_node(self, node: Node):
             The node to remove from the list.
         """
         self.nodes.remove(node)
+        log.info("[NODE-{}] Successfully removed Node".format(node.name))
```
Comment (Devoxin) [inline lavalink/nodemanager.py]: Nit: Lowercase `N` in `Node`

File: lavalink/playermanager.py [hunk 12]
```diff
@@ -1,7 +1,11 @@
+import logging
+
 from .node import Node
 from .models import BasePlayer
 from .exceptions import NodeException
 
+log = logging.getLogger('lavalink')
```
Comment (Devoxin) [inline lavalink/playermanager.py]: `__name__`, to remain consistent with `__init__.py`, unless there's a specific reason for using `lavalink`?

File: lavalink/playermanager.py [hunk 13]
```diff
@@ -41,6 +45,8 @@ def __iter__(self):
         if player.node and player.node.available:
             await player.node._send(op='destroy', guildId=player.guild_id)
 
+        log.info("[NODE-{}] Successfully destroyed its player".format(player.node.name))
```
Comment (Devoxin) [inline lavalink/playermanager.py]: Nit: `Successfully destroyed player for GuildId: {player.guild_id}`

File: lavalink/playermanager.py [hunk 14]
```diff
@@ -108,4 +114,5 @@ def create(self, guild_id: int, region: str = 'eu', endpoint: str = None, node:
         self.players[guild_id] = player = self.default_player(guild_id, node)
         node._original_players.append(player)
 
+        log.info("[NODE-{}] Successfully created a player".format(node.name))
```
Comment (Devoxin) [inline lavalink/playermanager.py]: Nit: `...for GuildId: {guild_id}`

File: lavalink/models.py (hunk 9 above)
Comment (Devoxin) [reply lavalink/models.py]: It does, but we're not using kwarg here

File: lavalink/client.py (hunk 7 above)
Comment (Devoxin) [reply lavalink/client.py]: Good point. Maybe remove support for non-async funcs. Can mark it as a breaking change

File: lavalink/client.py [hunk 15]
```diff
@@ -51,16 +51,19 @@ def __init__(self, user_id: int, shard_count: int = 1,
...
 
         self._session = aiohttp.ClientSession(
             connector=aiohttp.TCPConnector(loop=loop),
             timeout=aiohttp.ClientTimeout(total=30)
         )  # This session will be used for websocket and http requests.
 
     def add_event_hook(self, hook):
-        if hook not in self._event_hooks:
-            self._event_hooks.append(hook)
+        if hook not in self._event_hooks.get("Unknown Event"):
```
Comment (Devoxin) [inline lavalink/client.py]: If I understand this correctly, what if `Unknown Event` doesn't exist? You can't do `if hook in None`

File: lavalink/models.py [hunk 16]
```diff
@@ -223,7 +228,11 @@ def add(self, requester: int, track: dict, extra: dict = None, index: int = None
                 track = self.queue.pop(0)
 
         self.current = track
-        await self.node._send(op='play', guildId=self.guild_id, track=track.track, startTime=start_time)
+        if end_time is -1:
```
Comment (Devoxin) [inline lavalink/models.py]: `==`

File: lavalink/client.py (hunk 2 above)
Comment (Devoxin) [reply lavalink/client.py]: Wdym?

File: lavalink/client.py [hunk 17]
```diff
@@ -221,7 +230,7 @@ def decorator(func):
             The event to dispatch to the hooks.
         """
         try:
-            unknown_events = self._event_hooks.get('Unknown Events')
+            unknown_events = self._event_hooks.get(None)
```
Comment (Devoxin) [inline lavalink/client.py]: wat

File: lavalink/models.py [hunk 18]
```diff
@@ -229,8 +229,11 @@ def add(self, requester: int, track: dict, extra: dict = None, index: int = None
                 track = self.queue.pop(0)
 
         self.current = track
+
         if end_time is -1:
             end_time = track.duration
+        elif 0 > end_time > track.duration:
+            raise TypeError("end_time is either less than 0 or greater than the track's duration")
```
Comment (Devoxin) [inline lavalink/models.py]: single quotes

File: lavalink/client.py (hunk 15 above)
Comment (SoulSen) [reply lavalink/client.py]: Changed it in another commit

File: lavalink/models.py (hunk 9 above)
Comment (SoulSen) [reply lavalink/models.py]: So this is fine...?

File: lavalink/playermanager.py (hunk 12 above)
Comment (SoulSen) [reply lavalink/playermanager.py]: It was like that in `client.py`, want me to change it everywhere to that? I like that way since it's a bit more readable nothing major honestly.

File: lavalink/client.py (hunk 5 above)
Comment (SoulSen) [reply lavalink/client.py]: Must be two for loops or keep it like this if we are going to make the for loop async.

File: lavalink/client.py (hunk 7 above)
Comment (SoulSen) [reply lavalink/client.py]: Then `itertools.chain()` will not work.

File: lavalink/client.py (hunk 2 above)
Comment (Devoxin) [reply lavalink/client.py]: Oh, you mean if statement? `if not event_hooks`

File: lavalink/client.py (hunk 7 above)
Comment (Devoxin) [reply lavalink/client.py]: Can we not patch in our own version and put it in utils or something?
```py
def chain(*arrays):
  for a in arrays:
    for e in a:
      yield e
```

used like so:
```py
for hook in utils.chain(generic_hooks, targeted_hooks):
  await hook(event)
```

File: lavalink/models.py (hunk 9 above)
Comment (Devoxin) [reply lavalink/models.py]: I should think so

File: lavalink/playermanager.py (hunk 12 above)
Comment (Devoxin) [reply lavalink/playermanager.py]: Yeah, good point. Could you rename any `getLogger(__name__)` to `getLogger('lavalink')` then please?

File: lavalink/models.py [hunk 19]
```diff
@@ -220,7 +229,14 @@ def add(self, requester: int, track: dict, index: int = None):
                 track = self.queue.pop(0)
 
         self.current = track
-        await self.node._send(op='play', guildId=self.guild_id, track=track.track, startTime=start_time)
+
+        if end_time is -1:
+            end_time = track.duration
+        elif 0 > end_time > track.duration:
```
Comment (Zenrac) [inline lavalink/models.py]: You are handling error for `end_time` but what about `start_time` ?

File: lavalink/models.py (hunk 18 above)
Comment (Devoxin) [reply lavalink/models.py]: Yeah, `ValueError` would be better here

File: lavalink/models.py [hunk 20]
```diff
@@ -288,8 +304,8 @@ def add(self, requester: int, track: dict, index: int = None):
             band = value[0]
             gain = value[1]
 
-            if -1 > value[0] > 15:
-                continue
+            if not -1 < value[0] < 15:
+                raise InvalidBand('{} is a invalid band, must be 0-14'.format(band))
```
[the remark this answers is not in the document]
Comment (Devoxin) [reply lavalink/models.py]: Generally, user errors should be halted immediately and the issue should be made known to the user. I do see your point, but I believe this error has a more predictable behavior; you see an error so you expect execution of the function to stop. If it proceeds with execution, it might confuse the user. Additionally, the user of this library can catch these errors and perhaps forward them onto their end-users, so that if they implement EQ functionality and the end-user passes an invalid band, the message can be forwarded on to let the user know the provided input is invalid.

As for the grammar issue, `an invalid` are the correct words so this will need changing in that respect.

Also on another note, I can't remember if I mentioned it but I'm torn between whether this deserves its own exception class or if it should be an `IndexError`.

File: lavalink/models.py (hunk 16 above)
Comment (Devoxin) [reply lavalink/models.py]: I think modifying function parameter values to shave a few ns off processing times is probably pointless here, but I do agree that perhaps we shouldn't be sending stuff to Lavalink that doesn't need to be sent.

File: lavalink/models.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (SoulSen) [reply lavalink/models.py]: I think just using IndexError would be better.

File: lavalink/client.py (hunk 6 above)
Comment (SoulSen) [reply lavalink/client.py]: I need clearing up on this, on why you might want that.

**Comment by SoulSen** [comment]
- Breaking change now, all event hooks must be asynchronous. 
- `Client._dispatch_event()` now uses two asynchronous `for` loops to go through "generic" event hooks, and the registered/known event hooks. 
- `Play` command in `DefaultPlayer` now only sends the optional parameters it's given 
- Other than `log` being a local variable in some files, it's now `Client._logger`  
- `logging.getLogger()` takes `lavalink` and no longer `__name__`, helps with readability.

File: lavalink/client.py (hunk 5 above)
Comment (AndyTempel) [reply lavalink/client.py]: Wdym? Iterator has nothing to do with async

File: lavalink/client.py (hunk 6 above)
Comment (AndyTempel) [reply lavalink/client.py]: Adding it to queue instead of waiting for it to finish. If someones hook takes longer it could stall the ws loop since this func is awaited (if I'm not mistaken)

File: lavalink/client.py (hunk 5 above)
Comment (SoulSen) [reply lavalink/client.py]: Linter picks up that `itertools.chain()` is not a `AsyncIterable`, or something like that. Thats why in my latest commits I just created two async for loops, other than going through the hassle of doing a bunch of these things.

Comment (AndyTempel) [reply lavalink/client.py]: There's no need to have `async for` since lists you're putting in are not awaitable

File: lavalink/client.py (hunk 6 above)
Comment (AndyTempel) [reply lavalink/client.py]: Example use of async for would be when you're pulling pages of some list from some api.

File: lavalink/client.py (hunk 5 above)
Comment (SoulSen) [reply lavalink/client.py]: Oh okay, I'll fix it up tomorrow

File: lavalink/models.py [hunk 21]
```diff
@@ -228,14 +228,18 @@ def add(self, requester: int, track: dict, extra: dict = None, index: int = None
                 track = self.queue.pop(0)
 
         self.current = track
+        options = {}
+
+        if 0 > end_time > track.duration or 0 > start_time > track.duration:
+            raise ValueError("end_time or start_time is either less than 0 or greater than the track's duration")
+        elif end_time == 0:
```
Comment (Zenrac) [inline lavalink/models.py]: It doesn't make sense to set a value to end_time here.

Comment (Zenrac) [reply lavalink/models.py]: Moreover this whole change is kindda weird, it doesn't allow to have noreplace, endtime and/or start time at the same time ? You're not checking if start time is valid as well. See my old example if it can help you

File: lavalink/models.py [hunk 22]
```diff
@@ -230,12 +230,12 @@ def add(self, requester: int, track: dict, extra: dict = None, index: int = None
         self.current = track
         options = {}
 
-        if 0 > end_time > track.duration or 0 > start_time > track.duration:
-            raise ValueError("end_time or start_time is either less than 0 or greater than the track's duration")
-        elif end_time == 0:
+        if end_time == 0:
             options['endTime'] = track.duration
```
Comment (Devoxin) [inline lavalink/models.py]: This line & the one above can be omitted, as Lavalink (by default) will play until the end of a track

File: lavalink/models.py [hunk 23]
```diff
@@ -230,12 +230,12 @@ def add(self, requester: int, track: dict, extra: dict = None, index: int = None
...
         elif no_replace:
             options['noReplace'] = no_replace
```
Comment (Devoxin) [inline lavalink/models.py]: Perhaps this shouldn't be inside an if statement?

File: lavalink/models.py [hunk 24]
```diff
@@ -230,12 +230,12 @@ def add(self, requester: int, track: dict, extra: dict = None, index: int = None
...
+        elif 0 > end_time > track.duration or 0 > start_time > track.duration:
```
Comment (Devoxin) [inline lavalink/models.py]: start_time should have its own if statement, along with checks

File: lavalink/models.py [hunk 25]
```diff
@@ -230,12 +230,12 @@ def add(self, requester: int, track: dict, extra: dict = None, index: int = None
...
+            raise ValueError("end_time or start_time is either less than 0 or greater than the track's duration")
```
Comment (Devoxin) [inline lavalink/models.py]: single quotes

File: lavalink/models.py (hunk 21 above)
Comment (SoulSen) [reply lavalink/models.py]: Oh oops, they should've been separate if statements, will fix

**Comment by SoulSen** [comment]
### New Changes

- Queue system for dispatching event hooks. 
- Exceptions of event hooks are also raised now and not just logged. 
- Checks of `startTime` and `endTime` were changed.

After running tests, after the latest commit: 
```
-- flake8 test --
OK
-- pylint test --
OK
```

**Comment by SoulSen** [comment]
Oh wait, I just realized their is something wrong with the changes I made with `DefaultPlayer.play()`.

**Comment by SoulSen** [comment]
### Fixes 

- Fixed `start_time` and `end_time` parameter not being sent when modified.

File: lavalink/models.py [hunk 26]
```diff
@@ -220,7 +228,26 @@ def add(self, requester: int, track: dict, index: int = None):
                 track = self.queue.pop(0)
 
         self.current = track
-        await self.node._send(op='play', guildId=self.guild_id, track=track.track, startTime=start_time)
+        options = {}
+
+        if not 0 > start_time > track.duration and start_time:
```
Comment (Zenrac) [inline lavalink/models.py]: I don't really like the way you use `if elif pass else`.. can you use a more elegant if-else statement ? 
Such as : 
```py
if start_time:
    if 0 > start_time > track.duration:
        raise ValueError('start_time is either less than 0 or greater than the track\'s duration')
    options['startTime'] = start_time
```

File: lavalink/models.py [hunk 27]
```diff
@@ -220,7 +228,26 @@ def add(self, requester: int, track: dict, index: int = None):
...
+            options['startTime'] = start_time
+        elif start_time == 0:
+            pass
+        else:
+            raise ValueError('start_time is either less than 0 or greater than the track\'s duration')
+
+        if not 0 > end_time > track.duration and end_time:
```
Comment (Zenrac) [inline lavalink/models.py]: Same thing here

File: lavalink/client.py (hunk 2 above)
Comment (Zenrac) [reply lavalink/client.py]: Still not improved here @SoulSen :^)

Comment (SoulSen) [reply lavalink/client.py]: Not sure of any performant way though.

Comment (Zenrac) [reply lavalink/client.py]: `if not event_hooks` instead of `if event_hooks is None` :^)

File: lavalink/client.py [hunk 28]
```diff
@@ -224,16 +225,16 @@ def decorator(func):
         :param event:
             The event to dispatch to the hooks.
         """
-        generic_events = self._event_hooks.get('Generic') or []
-        registered_events = self._event_hooks.get(event) or []
+        generic_hooks = self._event_hooks.get('Generic') or []
```
Comment (Devoxin) [inline lavalink/client.py]: `.get('Generic', [])`

File: lavalink/client.py [hunk 29]
```diff
@@ -224,16 +225,16 @@ def decorator(func):
...
+        targeted_hooks = self._event_hooks.get(event) or []
```
Comment (Devoxin) [inline lavalink/client.py]: `.get(event, [])`

File: lavalink/client.py [hunk 30]
```diff
@@ -71,7 +71,7 @@ def add_event_hook(self, hook, event: Event = None):
         event_hooks = self._event_hooks.get(event)
 
         if hook not in event_hooks:
-            if event_hooks is None:
+            if not event_hooks:
                 self._event_hooks["Generic"] = [hook]
```
Comment (Devoxin) [inline lavalink/client.py]: `'Generic'`

File: lavalink/client.py [hunk 31]
```diff
@@ -211,14 +211,15 @@ def add_node(self, host: str, port: int, password: str, region: str,
         else:
             return
 
-    def on(self, event: Event = "Generic"):
+    def on(self, event: Event = 'Generic'):
```
Comment (Devoxin) [inline lavalink/client.py]: Not sure how I feel about *wanting* an `Event` but receiving `str` instead (mixed types). Perhaps this can be `None` (to indicate "No specific event) which defaults to Generic within the code?

File: lavalink/client.py [hunk 32]
```diff
@@ -58,13 +58,22 @@ def __init__(self, user_id: int, shard_count: int = 1,
             timeout=aiohttp.ClientTimeout(total=30)
         )  # This session will be used for websocket and http requests.
 
-    def add_event_hook(self, hook):
-        event_hooks = self._event_hooks.get('Unknown Event')
+    def add_event_hook(self, hook, event: Event = None):
+        """
+
+        :param hook:
+            The hook to be added, that will be dispatched when an event is dispatched.
```
Comment (Devoxin) [inline lavalink/client.py]: Nit: `The hook to register for the given event type.`

File: lavalink/client.py [hunk 33]
```diff
@@ -69,9 +69,10 @@ def add_event_hook(self, hook, event: Event = None):
             dispatched.
         """
         event_hooks = self._event_hooks.get(event)
+
         if hook not in event_hooks:
             if event_hooks is None:
-                self._event_hooks[event] = [hook]
+                self._event_hooks["Generic"] = [hook]
```
Comment (Devoxin) [inline lavalink/client.py]: FWIW this line is also bugged in the sense that if an Event type is specified but the array is empty, then the hook will be registered for all events rather than the specific type.

File: lavalink/client.py [hunk 34]
```diff
@@ -69,9 +69,10 @@ def add_event_hook(self, hook, event: Event = None):
             dispatched.
         """
         event_hooks = self._event_hooks.get(event)
+
         if hook not in event_hooks:
```
Comment (Devoxin) [inline lavalink/client.py]: If `event_hooks` is None, this will throw:

File: lavalink/client.py [hunk 35]
```diff
@@ -47,35 +48,14 @@ def __init__(self, user_id: int, shard_count: int = 1,
         self._shard_count = str(shard_count)
         self._loop = loop or asyncio.get_event_loop()
         self.node_manager = NodeManager(self, regions)
-        self.players = PlayerManager(self, player)
+        self.player_manager = PlayerManager(self, player)
```
Comment (Zenrac) [inline lavalink/client.py]: why

Comment (SoulSen) [reply lavalink/client.py]: Do you really want `Client.players.players`

Comment (Zenrac) [reply lavalink/client.py]: Well it's a breaking change then

Comment (SoulSen) [reply lavalink/client.py]: I need to change the example, I forgot to do that.

Comment (Devoxin) [reply lavalink/client.py]: I'm okay with breaking changes as the hook thing is a breaking change. I'll just bump the minor version. (I should follow semver (i.e. bump major version) but I really do not feel like this warrants `Lavalink.py v4`)

Comment (SoulSen) [reply lavalink/client.py]: What about version `3.1.0`

File: lavalink/__init__.py [hunk 36]
```diff
@@ -36,3 +38,39 @@ def enable_debug_logging():
     log.addHandler(handler)
 
     log.setLevel(logging.DEBUG)
+
+
+def _internal_add_event_hook(hook, event: Event = None):
```
Comment (Devoxin) [inline lavalink/__init__.py]: This function should probably do some kind of type checking. I.e., `if event is not None and not isinstance(event, lavalink.events.Event)` or something.

Comment (SoulSen) [reply lavalink/__init__.py]: What should it raise?

Comment (Devoxin) [reply lavalink/__init__.py]: TypeError('Event parameter is not of type `Event`')

File: lavalink/client.py (hunk 35 above)
Comment (Devoxin) [reply lavalink/client.py]: That's what I meant by minor version lol

File: lavalink/events.py [hunk 37]
```diff
@@ -11,6 +11,8 @@ class QueueEndEvent(Event):
     player: BasePlayer
         The player that has no more songs in queue.
     """
+    __slots__ = 'player'
```
Comment (Devoxin) [inline lavalink/events.py]: Doesn't `__slots__` need to be a list/tuple?

File: lavalink/events.py [hunk 38]
```diff
@@ -135,6 +149,8 @@ class NodeConnectedEvent(Event):
     node: Node
         The node that was successfully connected to.
     """
+    __slots__ = 'node'
```
Comment (Devoxin) [inline lavalink/events.py]: See above

File: lavalink/__init__.py [hunk 39]
```diff
@@ -49,6 +50,12 @@ def add_event_hook(hook, event: Event = None):
         The event the hook belongs to. This will dispatch when that specific event is
         dispatched.
     """
+    if event is not None or not isinstance(event, Event):
```
Comment (Devoxin) [inline lavalink/__init__.py]: This will throw if event is None.

File: lavalink/__init__.py [hunk 40]
```diff
@@ -49,6 +50,12 @@ def add_event_hook(hook, event: Event = None):
...
+        raise TypeError('Event parameter is not of type Event or None')
+
+    if not callable(hook) or inspect.iscoroutinefunction(hook):
+        raise TypeError('Hook is not callable or a coroutine')
```
Comment (Devoxin) [inline lavalink/__init__.py]: Missing `not` after `or`? I thought this change meant non-coroutine hooks couldn't be used?

Comment (SoulSen) [reply lavalink/__init__.py]: Yeah forgot it, my bad.

File: lavalink/events.py (hunk 37 above)
Comment (SoulSen) [reply lavalink/events.py]: Linter tells me to remove redundant `()`, so I'm gonna guess it doesn't

File: lavalink/__init__.py [hunk 41]
```diff
@@ -1,38 +1,84 @@
...
+        The hook to register for the given event type.
+        If `event` parameter is left empty, then it will run when any event is dispatched.
+    :param event:
+        The event the hook belongs to. This will dispatch when that specific event is
+        dispatched.
+    """
+    if event is not None and not isinstance(event, Event):
+        raise TypeError('Event parameter is not of type Event or None')
+
+    if not callable(hook) or not inspect.iscoroutinefunction(hook):
```
Comment (Zenrac) [inline lavalink/__init__.py]: Missing parentheses ?
image

Comment (SoulSen) [reply lavalink/__init__.py]: For which conditional?

Comment (Zenrac) [reply lavalink/__init__.py]: `if not callable(hook) or not inspect.iscoroutinefunction(hook):`

Comment (Zenrac) [reply lavalink/__init__.py]: Oh wat, there are 2 "not". nvm then

File: lavalink/models.py [hunk 42]
```diff
@@ -15,17 +15,24 @@ class TrackNotBuilt(Exception):
...
+                 'extra')
 
-    def __init__(self, requester, **kwargs):
+    def __init__(self, requester):
         self.requester = requester
-        self.preferences = kwargs
 
     @classmethod
-    def build(cls, track, requester, **kwargs):
+    def build(cls, track, requester, extra: dict = None):
```
[the remark this answers is not in the document]
Comment (Devoxin) [reply lavalink/models.py]: I believe he said that the linter picks up on this and doesn't like it
